masonry/src/widget/widget.rs

```diff
@@ -303,35 +324,36 @@ pub trait Widget: AsAny {
     }
 }
 
-pub(crate) fn get_child_at_pos<'c>(
-    widget: &(impl Widget + ?Sized),
-    ctx: QueryCtx<'c>,
+pub(crate) fn find_widget_at_pos<'c>(
+    widget: &WidgetRef<'c, dyn Widget>,
     pos: Point,
 ) -> Option<WidgetRef<'c, dyn Widget>> {
-    let relative_pos = pos - ctx.window_origin().to_vec2();
-    if !ctx
-        .clip_path()
-        .map_or(true, |clip| clip.contains(relative_pos))
-    {
-        return None;
-    }
-
-    // Assumes `Self::children_ids` is in increasing "z-order", picking the last child in case
-    // of overlapping children.
-    for child_id in widget.children_ids().iter().rev() {
-        let child = ctx.get(*child_id);
-
-        // The position must be inside the child's layout and inside the child's clip path (if
-        // any).
-        if !child.ctx().is_stashed()
-            && child.ctx().accepts_pointer_interaction()
-            && child.ctx().window_layout_rect().contains(pos)
+    if widget.ctx.widget_state.bounding_rect.contains(pos) {
+        let local_pos = widget.ctx().widget_state.window_transform.inverse() * pos;
+
+        if Some(false) == widget.ctx.clip_path().map(|clip| clip.contains(local_pos)) {
+            return None;
+        }
+
+        // Assumes `Self::children_ids` is in increasing "z-order", picking the last child in case
+        // of overlapping children.
+        for child_id in widget.children_ids().iter().rev() {
+            let child_ref = widget.ctx.get(*child_id);
+            if let Some(child) = child_ref.widget.find_widget_at_pos(child_ref.ctx, pos) {
+                return Some(child);
+            }
+        }
+        if !widget.ctx.is_stashed()
+            && widget.ctx.accepts_pointer_interaction()
+            && widget.ctx.size().to_rect().contains(local_pos)
         {
-            return Some(child);
+            Some(*widget)
+        } else {
+            None
         }
+    } else {
+        None
     }
```

Comment:
This is a core change in the PR, and it's one with a lot of semantic importance, so I'm giving it extra scrutiny. Some notes:

- This code has a lot of rightward drift that could be avoided with early returns.
- Again the difference between "declared layout rect" and "bounding rect of self and children" shows up. I think your code handles it correctly, but it does so implicitly. I think that distinction should be documented in a few places and given special attention in this function.
- I think `is_stashed` should be considered before the children. A stashed widget is never going to have un-stashed children. `accepts_pointer_interaction` is probably fine here, if we want to behave like `pointer-events: none` in the DOM, but that should be documented.

Reply:
> This code has a lot of rightward drift that could be avoided with early returns.

More than you have listed in the other points? I only see that `is_stashed` can be moved up (I've fixed this in the latest commit).

> Again the difference between "declared layout rect" and "bounding rect of self and children" shows up. I think your code handles it correctly, but it does so implicitly. I think that distinction should be documented in a few places and given special attention in this function.

Does it? the bounding rect is only used to filter out early as optimization, nothing else really. Otherwise I think it's documented what `bounding_rect` is, do you have more concrete suggestions?

> accepts_pointer_interaction is probably fine here, if we want to behave like pointer-events: none in the DOM, but that should be documented.

I agree that things should be more documented (and I've documented this in `Widget::find_widget_at_pos` now), but it's basically the same behavior as before with `get_child_at_pos` (which wasn't documented).